File: tests/test_webui_object_topology_contract.py
"""计划书 §11.2：WebUI 语义拓扑投影的契约测试。

前端契约用 headless node 跑 search_map.js 暴露的纯函数
``window.TopologyLayout.objectOnly``；后端契约直接检查
``SemanticNavigationGraph`` 的投影同时满足同一份约束。
"""

from __future__ import annotations

import shutil
import subprocess
from pathlib import Path

import pytest

from app.spatial.semantic_navigation_graph import SemanticNavigationGraph

NODE = shutil.which("node")
JS_TEST = Path(__file__).with_name("test_webui_object_topology_contract.js")
SEARCH_MAP = Path(__file__).resolve().parents[1] / "app/manual_web_demo/static/search_map.js"


@pytest.mark.skipif(NODE is None, reason="node.js not available")
def test_frontend_projection_contract():
    result = subprocess.run(
        [NODE, str(JS_TEST)], capture_output=True, text=True, timeout=60,
    )
    assert result.returncode == 0, (
        f"node topology contract test failed:\n{result.stdout}\n{result.stderr}"
    )
    assert "ALL JS TOPOLOGY CONTRACT TESTS PASSED" in result.stdout


def test_render_never_prefers_the_whole_navigation_graph():
    text = SEARCH_MAP.read_text(encoding="utf-8")
    assert "var topology = graph && graph.object_topology ? graph.object_topology : null;" in text
    # 旧的"先整张导航图，空了才退回 object_topology"逻辑必须彻底消失。
    assert "topology = (graph && graph.object_topology) || null;" not in text
    assert "var topology = graph || null;" not in text


def _graph_with_place_frontier_and_objects() -> SemanticNavigationGraph:
    graph = SemanticNavigationGraph(relation_confirm_min_observations=1)
    for index, observation in enumerate(("bundle_1", "bundle_2"), start=1):
        graph.update_observation(
            observation_id=observation,
            heading_sector=index - 1,
            scene_objects=[
                {"id": f"frame_{index}_1", "label": "办公桌",
                 "map_xyz": [1.0, 0.0, 0.0], "confidence": 0.9},
                {"id": f"frame_{index}_2", "label": "白色垃圾桶",
                 "map_xyz": [1.3, 0.5, 0.0], "confidence": 0.9},
            ],
            scene_relations=[{"subject_id": f"frame_{index}_1",
                              "object_id": f"frame_{index}_2",
                              "relation": "near", "confidence": 0.8}],
            pose={"x": 0.0, "y": 0.0, "yaw_rad": 0.0}, timestamp=float(index),
        )
    return graph


def _graph_with_p1_f01_and_two_objects() -> SemanticNavigationGraph:
    """计划书 §11.2 的完整导航图：P1 + F01 + obj_001 + obj_002。

    观测只覆盖 sector 1、2，所以 sector 0 留下的 frontier 短标签正好是 F01。
    """
    graph = SemanticNavigationGraph(relation_confirm_min_observations=1)
    for index, sector in enumerate((1, 2), start=1):
        graph.update_observation(
            observation_id=f"bundle_{index}", heading_sector=sector,
            scene_objects=[
                {"id": f"frame_{index}_1", "label": "办公桌",
                 "map_xyz": [1.0, 0.0, 0.0], "confidence": 0.9},
                {"id": f"frame_{index}_2", "label": "白色垃圾桶",
                 "map_xyz": [1.3, 0.5, 0.0], "confidence": 0.9},
            ],
            scene_relations=[{"subject_id": f"frame_{index}_1",
                              "object_id": f"frame_{index}_2",
                              "relation": "near", "confidence": 0.8}],
            pose={"x": 0.0, "y": 0.0, "yaw_rad": 0.0}, timestamp=float(index),
        )
    return graph


def test_projection_keeps_only_obj_001_and_obj_002():
    graph = _graph_with_p1_f01_and_two_objects()
    full = graph.to_dict()
    internal_ids = {node["node_id"] for node in full["nodes"]}
    internal_labels = {str(node.get("label")) for node in full["nodes"]}
    # 内部导航图必须同时含 P1 和短标签 F01 的 frontier，规划器要用，不许删。
    assert "P1" in internal_ids
    assert "F01" in internal_labels
    assert {"obj_001", "obj_002"} <= internal_ids

    topology = full["object_topology"]
    node_ids = {node["node_id"] for node in topology["nodes"]}
    labels = {str(node.get("label")) for node in topology["nodes"]}
    assert node_ids == {"obj_001", "obj_002"}
    assert "P1" not in node_ids and "P1" not in labels
    assert "F01" not in node_ids and "F01" not in labels
    for edge in topology["edges"]:
        assert edge["from"] in node_ids and edge["to"] in node_ids
    relations = {edge["relation"] for edge in topology["edges"]}
    assert "OBSERVED_FROM" not in relations
    assert "FRONTIER_TO" not in relations


def test_backend_projection_excludes_place_and_frontier():
    graph = _graph_with_place_frontier_and_objects()
    full = graph.to_dict()
    node_types = {node["node_type"] for node in full["nodes"]}
    # 内部导航图仍然保留 Place（以及产生的 Frontier），规划器要用。
    assert "PLACE" in node_types

    topology = full["object_topology"]
    assert topology["schema_version"] == "semantic_object_topology_v1"
    object_ids = {node["node_id"] for node in topology["nodes"]}
    assert object_ids, "expected persistent object nodes"
    assert all(node["node_type"] == "OBJECT" for node in topology["nodes"])
    assert not any(node_id.startswith(("P", "F")) for node_id in object_ids)
    for edge in topology["edges"]:
        assert edge["from"] in object_ids and edge["to"] in object_ids
        assert edge["relation"] not in {"OBSERVED_FROM", "FRONTIER_TO", "MOVED_TO"}

File: tests/test_webui_object_topology_contract.js
#!/usr/bin/env node
"use strict";
/*
 * 拓扑契约测试。
 *
 * 构造同时含 P1、F01、obj_001、obj_002 的完整导航图（semantic_navigation_graph_v1），
 * 断言 WebUI 的语义拓扑投影保留对象节点 + PLACE 中转站及其挂接边（OBSERVED_FROM /
 * CONNECTED_TO），并剔除 FRONTIER 节点与 FRONTIER_TO / MOVED_TO 纯导航关系。
 * PLACE 必须留下：去掉它，不同视角看到的物体之间没有任何边，拓扑必然碎片化。
 *
 * Run with:  node tests/test_webui_object_topology_contract.js
 */
const fs = require("fs");
const path = require("path");
const vm = require("vm");

const searchMapPath = path.join(
  __dirname, "..", "app", "manual_web_demo", "static", "search_map.js");
const source = fs.readFileSync(searchMapPath, "utf8");

let failures = 0;
function assert(cond, msg) {
  if (!cond) { console.error("  FAIL: " + msg); failures += 1; return; }
  console.log("  ok: " + msg);
}
function assertEqual(a, b, msg) {
  if (a !== b) {
    console.error("  FAIL: " + msg + " (expected " + JSON.stringify(b) +
                  ", got " + JSON.stringify(a) + ")");
    failures += 1;
    return;
  }
  console.log("  ok: " + msg);
}

const window = {};
const sandbox = { window: window, document: undefined, console: console };
vm.createContext(sandbox);
vm.runInContext(source, sandbox, { filename: searchMapPath });

const objectOnly = window.TopologyLayout && window.TopologyLayout.objectOnly;
if (!objectOnly) {
  console.error("window.TopologyLayout.objectOnly not exposed - cannot test");
  process.exit(1);
}

// ---- 完整导航图：Place + Frontier + Object，含导航关系边 -------------------
const navigationGraph = {
  schema_version: "semantic_navigation_graph_v1",
  revision: 7,
  current_place_id: "P1",
  nodes: [
    { node_id: "P1", node_type: "PLACE", label: "P1", current: true },
    { node_id: "F01", node_type: "FRONTIER", label: "F01", status: "OPEN" },
    { node_id: "F02", node_type: "FRONTIER", label: "F02", status: "OPEN" },
    { node_id: "obj_001", node_type: "OBJECT", label: "办公桌",
      status: "CONFIRMED", observation_count: 4 },
    { node_id: "obj_002", node_type: "OBJECT", label: "白色垃圾桶",
      status: "CONFIRMED", observation_count: 2 },
  ],
  edges: [
    { from: "obj_001", to: "obj_002", relation: "near", status: "CONFIRMED" },
    { from: "P1", to: "obj_001", relation: "OBSERVED_FROM" },
    { from: "P1", to: "F01", relation: "FRONTIER_TO" },
    { from: "P1", to: "F02", relation: "MOVED_TO" },
  ],
};

// 后端已经过滤好的对象投影（正常路径）。
const objectTopology = {
  schema_version: "semantic_object_topology_v1",
  revision: 7,
  nodes: [
    { node_id: "obj_001", node_type: "OBJECT", label: "办公桌",
      status: "CONFIRMED", observation_count: 4 },
    { node_id: "obj_002", node_type: "OBJECT", label: "白色垃圾桶",
      status: "CONFIRMED", observation_count: 2 },
  ],
  edges: [
    { from: "obj_001", to: "obj_002", relation: "near", status: "CONFIRMED" },
  ],
  stats: { node_count: 2, edge_count: 1 },
};

function ids(topology) {
  return (topology.nodes || []).map(function (n) { return n.node_id; }).sort();
}

console.log("case 1: backend object_topology passes through unchanged");
const projected = objectOnly(objectTopology);
assertEqual(JSON.stringify(ids(projected)),
            JSON.stringify(["obj_001", "obj_002"]), "nodes = obj_001, obj_002");
assertEqual(projected.edges.length, 1, "one object relation edge survives");
assertEqual(projected.edges[0].relation, "near", "relation is near");
assertEqual(projected.schema_version, "semantic_object_topology_v1",
            "schema_version preserved");

console.log("case 2: a leaked full navigation graph keeps objects + place, drops frontiers");
const guarded = objectOnly(navigationGraph);
const guardedIds = ids(guarded);
// 排序后大写 P 的 ASCII 小于小写 o，所以 P1 在最前面。
assertEqual(JSON.stringify(guardedIds),
            JSON.stringify(["P1", "obj_001", "obj_002"]),
            "nodes = P1, obj_001, obj_002");
assert(guardedIds.indexOf("P1") >= 0, "P1 kept as the cross-view hub");
assert(guardedIds.indexOf("F01") < 0, "F01 absent");
assert(guardedIds.indexOf("F02") < 0, "F02 absent");
const relations = guarded.edges.map(function (e) { return e.relation; });
assert(relations.indexOf("OBSERVED_FROM") >= 0, "OBSERVED_FROM kept");
assert(relations.indexOf("FRONTIER_TO") < 0, "FRONTIER_TO absent");
assert(relations.indexOf("MOVED_TO") < 0, "MOVED_TO absent");
assertEqual(guarded.edges.length, 2, "near + OBSERVED_FROM remain");
guarded.edges.forEach(function (edge) {
  assert(guardedIds.indexOf(edge.from) >= 0 && guardedIds.indexOf(edge.to) >= 0,
         "edge endpoints subset of kept nodes: " + edge.from + "->" + edge.to);
});

console.log("case 3: node type comes from node_type, never from the id prefix");
const spoofed = objectOnly({
  nodes: [
    { node_id: "obj_bad", node_type: "FRONTIER", label: "伪装成物体的前沿" },
    { node_id: "P9", node_type: "OBJECT", label: "ID 不带前缀的真物体" },
  ],
  edges: [{ from: "obj_bad", to: "P9", relation: "near" }],
});
assertEqual(JSON.stringify(ids(spoofed)), JSON.stringify(["P9"]),
            "obj_ prefixed FRONTIER dropped, OBJECT kept regardless of id");
assertEqual(spoofed.edges.length, 0, "edge with a dropped endpoint is removed");

console.log("case 4: isolated objects are shown, no fallback to the nav graph");
const isolated = objectOnly({
  nodes: [{ node_id: "obj_001", node_type: "OBJECT", label: "办公桌" }],
  edges: [],
});
assertEqual(isolated.nodes.length, 1, "isolated object node kept");
assertEqual(isolated.edges.length, 0, "no relations invented");
assertEqual(objectOnly(null), null, "missing topology stays null (empty state)");
assertEqual(objectOnly({}), null, "topology without nodes stays null");

if (failures) {
  console.error("\n" + failures + " JS TOPOLOGY CONTRACT TEST(S) FAILED");
  process.exit(1);
}
console.log("\nALL JS TOPOLOGY CONTRACT TESTS PASSED");
